[MRG] Fix decoding of JPEG compressed color images with Photometric Interpretation "RGB"

hackermd opened the pull request:
The Pixel Data element of DICOM image instances may contain JPEG bitstreams that were not converted to YCbCr color space upon compression. In these cases the value of `PhotometricInterpretation` is `"RGB"` rather than `"YBR_FULL_422"`.

JPEG is color blind, i.e. the JPEG bitstream may not contain any information about any color transformation, and Pillow's `JpegImagePlugin` assumes that images have been transformed into YCbCr space before compression. As a consequence, the `pillow_handler` doesn't decode images correctly, which were compressed directly in RGB color space (see https://github.com/python-pillow/Pillow/issues/3952).  

This PR fixes the color "mode" of the `Pillow.Image` object by providing additional color information to the decoder and thereby ensure that images are decoded correctly.

pep8speaks commented:
Hello @hackermd! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pydicom/pixel_data_handlers/pillow_handler.py`](https://github.com/pydicom/pydicom/blob/97e222fc15069835663e69c9987f6fc685ffab69/pydicom/pixel_data_handlers/pillow_handler.py):

> [Line 83:80](https://github.com/pydicom/pydicom/blob/97e222fc15069835663e69c9987f6fc685ffab69/pydicom/pixel_data_handlers/pillow_handler.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 95:13](https://github.com/pydicom/pydicom/blob/97e222fc15069835663e69c9987f6fc685ffab69/pydicom/pixel_data_handlers/pillow_handler.py#L95): [E129](https://duckduckgo.com/?q=pep8%20E129) visually indented line with same indent as next logical line

hackermd commented:
I noticed that the `test_PI_RGB` test now fails. Are we sure that this test case has been correct in the first place?

darcymason commented:
> I noticed that the `test_PI_RGB` test now fails. Are we sure that this test case has been correct in the first place

I'm not sure ... @rhaxton might know ... from quick check it may come from #521 but I didn't really follow the history fully.

rhaxton commented:
The test_PI_RGB tests just test that the reading of lossy compressed jpegs behaves the same what that using dcmtk's dcmcjpeg and dcmdjpeg behave.

The filenames being tested include the command line options passed to dcmcjpeg (version 3.6.1 as indicated in the file metadata).  For example:

`SC_rgb_dcmtk_+eb+cy+np.dcm`

was created by running:

`dcmcjpeg +eb +cy +np SC_rgb.dcm SC_rgb_dcmtk_+eb+cy+np.dcm`

and the expected result was created by running:

`dcmdjpeg SC_rgb_dcmtk_+eb+cy+np.dcm SC_rgb_dcmtk_ebcynp_dcmd.dcm`

These tests only assert that we decode these files the same way dcmtk decodes them.

as noted in the discussion of #521, the gdcm and dcmtk command line tools disagreed in some of these cases.  The dcmtk results "looked" better, so that is what is in the test (and what is supported).

If these tests are failing, then the pillow handler is not replicating dcmtk's dcmdjpeg command line tool for that file.  Whether it is correct or not is an open question.  I would like to note that quite a lot of lossy compressed color jpeg DICOM files have the PI tag set to "RGB" incorrectly, and we should have some way to support them....

hackermd commented:
> I would like to note that quite a lot of lossy compressed color jpeg DICOM files have the PI tag set to "RGB" incorrectly, and we should have some way to support them....

I think you are right about the incorrectly set `"RGB"`. If the set the value of `PhotometricInterpretation` of the data set contained in file `SC_rgb_dcmtk_+eb+cr.dcm` to `"YBR_FULL_422"`, then the `pillow_handler.py` implementation in this PR decodes the pixel data correctly.  

I really don't like the idea of supporting invalid files, since this will lead to unexpected behavior down the line because users assume that files are valid and are then surprised if a different application fails to handle them.
If a file is invalid, the pixel handler should not try to provide a workaround IMO, but rather fail by raising an appropriate exception. I updated the test case and removed the invalid file (ab9c3fce084fae3240c63b00d0aa4bce8615b670).

hackermd commented:
@darcymason @rhaxton Are there any objections against merging this PR?

scaramallion commented:
There's no test coverage of the [actual correction](https://codecov.io/gh/pydicom/pydicom/src/36e1306adddb9b48b4f898426a36fe3c9d49364e/pydicom/pixel_data_handlers/pillow_handler.py).

scaramallion commented:
Why did you remove tests from the GDCM unit tests if this only affects Pillow?

scaramallion commented on pydicom/pixel_data_handlers/pillow_handler.py:
```diff
@@ -80,6 +80,33 @@ def should_change_PhotometricInterpretation_to_RGB(dicom_dataset):
+        # space prior to compression, setting the value of "mode" to YCbCr
+        # signals Pillow to not apply any color transformation upon
+        # decompression.
+        if (transfer_syntax in PillowJPEGTransferSyntaxes and
```
This could just be `if transfer_syntax.is_compressed ...` since `get_pixeldata()` already checks that its a valid transfer syntax.

scaramallion commented on pydicom/pixel_data_handlers/pillow_handler.py:
```diff
@@ -80,6 +80,33 @@ def should_change_PhotometricInterpretation_to_RGB(dicom_dataset):
 
 
+def _decompress_single_frame(data, transfer_syntax, photometric_interpretation):
+    fio = io.BytesIO(data)
```
Can you add a docstring please?

hackermd replied:
Addressed with 7f081202a48caad0c2b8c5061e95f034e4b6d0ed

hackermd commented:
@scaramallion I added a test case via 9cb6fc26fa896825b9fcb3acdb7fc2205f9efec4. It would great if this bug fix could make it into the next release.

scaramallion commented:
The test file you deleted goes from working OK in current `master` to raising an IOError due to a broken image stream. Why was it removed?

Also, the two test datasets you added aren't based on the same image as the other SC* datasets, they should be renamed and their source and an explanation of what they contain be added to the README in the test_files directory.

hackermd commented:
@scaramallion I renamed the file and added a description to the README as requested. Please let me know if anything else is required for getting this merged. I think it's important to get this fix released.

Image without the fix (pydicom 2.0.0):

<img width="447" alt="Screen Shot 2020-11-27 at 5 10 46 PM" src="https://user-images.githubusercontent.com/6855297/100486853-e3063680-30d3-11eb-82bd-ffb9e8766777.png">

Image with fix (b88d9b6d29d6ab8c934608a7c291125129d7a770):

<img width="447" alt="Screen Shot 2020-11-27 at 5 12 10 PM" src="https://user-images.githubusercontent.com/6855297/100486860-ec8f9e80-30d3-11eb-844b-4efa427503ca.png">

hackermd commented:
@scaramallion I updated the branch. First tried to rebase, but gave up after rebasing for over half an hour. Ultimately, ended up merging current master into the branch.

Tests are passing. It would really be great if you could get this merged.

scaramallion commented:
Dang, I thought that would re-run the CI
